python/enthic/download_from_INPI.py

```diff
@@ -0,0 +1,92 @@
+# -*- coding: utf-8 -*-
+"""
+========================================
+Download daily zip files from INPI's FTP
+========================================
+
+Coding Rules:
+
+- Snake case for variables.
+- Only argument is configuration file.
+- No output or print, just log and files.
+"""
+
+from os.path import dirname, join, basename
+from os import remove
+import subprocess
+from json import load
+from ftplib import FTP_TLS
+from logging import info
+
+from enthic.extract_bundle import process_daily_zip_file
+
+
+################################################################################
+# READ CONFIGURATION
+with open(join(dirname(__file__), "configuration.json")) as json_configuration_file:
+    CONFIG = load(json_configuration_file)
+
+def add_data_to_database(file_path):
+    """
+        Add data contained in the given file into MySQL database
+        then delete temporary files if it's a success
+
+        :param file_path: daily INPI file to add to database
+    """
+    # extract data from xml files to csv files
+    process_daily_zip_file(file_path)
+
+    # Load new data into database
+    result = subprocess.run(["../sh/database-filling.sh", CONFIG["mySQL"]["enthic"]["password"]], stdout=subprocess.PIPE, check=True)
+    if result.stdout:
+        raise Exception("Following error while loading data from file "
+                        + file_path + " into database : \n"
+                        + result.stdout.decode('utf-8'))
+    else:
+        remove(join(CONFIG['outputPath'], CONFIG['tmpBundleFile']))
+        remove(join(CONFIG['outputPath'], CONFIG['identityFile']))
+        remove(join(CONFIG['outputPath'], CONFIG['metadataFile']))
+        remove(file_path)
+
+
+def exploreAndProcessFTPFolder(folderToExplore) :
+    """
+    Recursive function that explore FTP to download every daily zip files
+
+    :param folderToExplore: folder to explore
+    """
+    info("Exploring INPI's FTP folder " + folderToExplore)
+    inpi_user = CONFIG["INPI"]["user"]
+    inpi_password = CONFIG["INPI"]["password"]
+    ftp = FTP_TLS('opendata-rncs.inpi.fr')
+    ftp.login(user=inpi_user, passwd=inpi_password)
+    ftp.prot_p()
+    connexion = True
+    element_list = ftp.nlst(folderToExplore)
+    for element in element_list:
+        if element.endswith(".zip"):
+            localfile_path = join(CONFIG['inputPath'], basename(element))
+            localfile = open(localfile_path, 'wb')
+            info("Downloading file " + element + " into file " + localfile_path)
+            ftp.retrbinary("RETR " + element, localfile.write)
+            add_data_to_database(localfile_path)
+        elif element.endswith(".md5"):
+            continue
+        else:
+            if connexion:
+                ftp.quit()
+                connexion = False
+            exploreAndProcessFTPFolder(element)
+
+    if connexion:
+        ftp.quit()
```

Comment:
Is the connection boolean really useful, why not just quitting once at the end?

Reply:
It's a recursive function, so I close FTP connexion at 2 places in the function : at exit or before calling it again. But closing twice the connexion raise an exception. Globally, FTP connexion is fragile, that's why I'm reconnecting for each FTP folder. But we could add an issue to enhance the FTP connexion error handling to be more efficient.